### PR TITLE
TEST-198 Refactor JsonWebToken API in MP TCK Runner JWT Auth 

### PR DESCRIPTION
**Pull request by AlanRoth**
Moved TokenParser class to Payara Server from MP TCK, plus changes to make it compatible with #3799 which is mostly the same changes #3799 does to the Parse method.

Relies on #44 in MicroProfile-TCK-Runners
https://github.com/payara/MicroProfile-TCK-Runners/pull/44

### CODE REVIEW

**Comment by AlanRoth** [comment]
jenkins test please

File: appserver/payara-appserver-modules/microprofile/jwt-auth/src/main/java/fish/payara/microprofile/jwtauth/jwt/JwtTokenParser.java [hunk 1]
```diff
@@ -81,8 +84,17 @@ public JwtTokenParser() {
...
+        rawToken = bearerToken;
+        signedJWT = SignedJWT.parse(rawToken);
+        
+        if(!checkIsJWT(signedJWT.getHeader())){
+            throw new IllegalStateException("Not JWT");
+        }
+    }
+    
+    public JsonWebTokenImpl verify(String issuer, PublicKey publicKey) throws Exception {
+        SignedJWT signedJWT = SignedJWT.parse(rawToken);
```
Comment (cubastanley) [inline appserver/payara-appserver-modules/microprofile/jwt-auth/src/main/java/fish/payara/microprofile/jwtauth/jwt/JwtTokenParser.java]: Don't feel great about the ambiguity of this local variable having the same name as the signedJWT variable at line 76 - could it have a slightly different name?

Comment (AlanRoth) [reply appserver/payara-appserver-modules/microprofile/jwt-auth/src/main/java/fish/payara/microprofile/jwtauth/jwt/JwtTokenParser.java]: signedJWT2? I don't think its worth the change as #3799 is refactoring this class, I just did the minimum to make it function for now with the least amount of changes; #3799 is due for a merge shortly after this.

Comment (AlanRoth) [reply appserver/payara-appserver-modules/microprofile/jwt-auth/src/main/java/fish/payara/microprofile/jwtauth/jwt/JwtTokenParser.java]: That line can be removed and a null check added - as you should call parse before calling verify

**Comment by AlanRoth** [comment]
jenkins test please

**Comment by AlanRoth** [comment]
jenkins test please

File: appserver/payara-appserver-modules/microprofile/jwt-auth/src/main/java/fish/payara/microprofile/jwtauth/jwt/JwtTokenParser.java [hunk 2]
```diff
@@ -81,8 +84,19 @@ public JwtTokenParser() {
...
+        rawToken = bearerToken;
+        signedJWT = SignedJWT.parse(rawToken);
+        
+        if(!checkIsJWT(signedJWT.getHeader())){
+            throw new IllegalStateException("Not JWT");
+        }
+    }
+    
+    public JsonWebTokenImpl verify(String issuer, PublicKey publicKey) throws Exception {
+        if(signedJWT == null){
```
Comment (MattGill98) [inline appserver/payara-appserver-modules/microprofile/jwt-auth/src/main/java/fish/payara/microprofile/jwtauth/jwt/JwtTokenParser.java]: This needs formatting
```suggestion
        if (signedJWT == null) {
```

File: appserver/payara-appserver-modules/microprofile/jwt-auth/src/main/java/fish/payara/microprofile/jwtauth/jwt/JwtTokenParser.java [hunk 3]
```diff
@@ -81,8 +84,19 @@ public JwtTokenParser() {
         this(Optional.empty(), Optional.empty());
     }
     
-    public JsonWebTokenImpl parse(String bearerToken, String issuer, PublicKey publicKey) throws Exception {
-        SignedJWT signedJWT = SignedJWT.parse(bearerToken);
+    public void parse(String bearerToken) throws Exception {
+        rawToken = bearerToken;
+        signedJWT = SignedJWT.parse(rawToken);
+        
+        if(!checkIsJWT(signedJWT.getHeader())){
```
Comment (MattGill98) [inline appserver/payara-appserver-modules/microprofile/jwt-auth/src/main/java/fish/payara/microprofile/jwtauth/jwt/JwtTokenParser.java]: Needs formatting
```suggestion
        if (!checkIsJWT(signedJWT.getHeader())) {
```

File: appserver/payara-appserver-modules/microprofile/jwt-auth/src/main/java/fish/payara/microprofile/jwtauth/jwt/JwtTokenParser.java [hunk 4]
```diff
@@ -72,6 +72,9 @@
     private final boolean enableNamespacedClaims;
     private final Optional<String> customNamespace;
 
+    private String rawToken;
+    private SignedJWT signedJWT;
```
Comment (MattGill98) [inline appserver/payara-appserver-modules/microprofile/jwt-auth/src/main/java/fish/payara/microprofile/jwtauth/jwt/JwtTokenParser.java]: It's typically bad practice to have two class variables depending directly on each other. Is it possible to make the signed JWT a local variable without breaking the TCK or introducing another problem?

Comment (AlanRoth) [reply appserver/payara-appserver-modules/microprofile/jwt-auth/src/main/java/fish/payara/microprofile/jwtauth/jwt/JwtTokenParser.java]: Shouldn't be too much trouble, I'll cook something up.

**Comment by AlanRoth** [comment]
> 
> 
> Agreed with Matt's requests, otherwise I'm happy

I agree too, and I had a chat with Matt and #3799 will revert any changes back to using those two variables, it wouldn't be too bad of a shout on pointing it out on #3799

**Comment by AlanRoth** [comment]
jenkins test please

**Pull request merged**